ssz_serialization/proofs.nim

```diff
@@ -252,6 +257,106 @@ func verify_merkle_multiproof*(
   if calc.isErr: return false
   calc.get == root
 
+func verify_merkle_multiproof*(
+    leaves: openArray[Digest],
+    proof: openArray[Digest],
+    indices: static openArray[GeneralizedIndex],
+    root: Digest): bool =
+  const helper_indices = get_helper_indices(indices)
+  let calc = calculate_multi_merkle_root(leaves, proof, indices, helper_indices)
+  if calc.isErr: return false
+  calc.get == root
+
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/tests/core/pyspec/eth2spec/test/helpers/merkle.py#L4-L21
+func build_proof*(
+    anchor: auto,
+    indices: openArray[GeneralizedIndex],
+    helper_indices: openArray[GeneralizedIndex],
+    proof: var openArray[Digest]): Result[void, string] =
+  doAssert proof.len == helper_indices.len
+  ? check_multiproof_acceptable(indices)
+  hash_tree_root(anchor, helper_indices, proof)
+
+func build_proof*(
+    anchor: auto,
+    indices: openArray[GeneralizedIndex],
+    proof: var openArray[Digest]): Result[void, string] =
+  ? check_multiproof_acceptable(indices)
+  let helper_indices = get_helper_indices(indices)
+  doAssert proof.len == helper_indices.len
+  hash_tree_root(anchor, helper_indices, proof)
+
+func build_proof*(
+    anchor: auto,
+    indices: static openArray[GeneralizedIndex],
+    proof: var openArray[Digest]): Result[void, string] =
```

Comment:
Is there still a reason for having the function signatures with `proof: var openArray[Digest]` and a returned void Result (and the same counts for the `root_tree_hash` counterparts)? 

When there is already the need for `Result[void, string]` (or boolean for that matter), perhaps that type of return should be only provided, considering it is a safer practice (caller must verify the result to access the proofs).

Reply:
I guess it could save some copying because the result can be directly put into the destination, but semantically these two are the same.
```
var foo: SomeObject
? state.build_proof(5.GeneralizedIndex, foo.proof)
? foo.proof = state.build_proof(5.GeneralizedIndex)
```